### PR TITLE
Augmentation PR

### PR DESCRIPTION
**Pull request by sbaghba**
what remains:

- [x] working version of `augmentation_visualizer.py` that properly displays the augmentations being applied
  - use @mkutu changes
  - @mkutu push includes changes to the dataset class that will break other subtasks like train and the dataloader visualizer. These minor fixes also need to be performed. 
- [x] make sure the each augmentation is working. Turn off all other, and go one-by-one
- [x] Add "important" arguments for each transform if they're missing.
- [x] add descriptions/comments for complex augments
- [x] add post-augment normalization
- [x] adding the custom collate labels in `augmentation_visualizer.py` is low priority. This should be done last if all other items are complete, tested, and pushed.

### CODE REVIEW

**Comment by mkutu** [comment]
@sbaghba anytime we run mask_gen with any project name other than "test" we get a project directory that points to train even though we're running mask_gen. Please fix this.

**Comment by sbaghba** [comment]
> @sbaghba anytime we run mask_gen with any project name other than "test" we get a project directory that points to train even though we're running mask_gen. Please fix this.

I dont think this is a problem. the logging folder is named train in paths. that is essentially the train log folder as i intended. if you add new modes you should alter the directories accordingly. otherwise my logging directory which is projec_name/train will be used.

File: src/data/collate.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
+# src/data/collate.py
+
+import random
+import numpy as np
+import torch
+import torch.nn.functional as F
+from torch.utils.data._utils.collate import default_collate
+
+def mixup_collate(batch, p: float, alpha: float):
```
Comment (mkutu) [inline src/data/collate.py]: needs output type annotations for all functions

**Comment by mkutu** [comment]
1. wrap-up `someof` functionality so that we get "someof" the augmentations that are "enabled". Use one someof for geometric and one someof for pixel-based augments. [link](https://albumentations.ai/docs/2-core-concepts/pipelines/#someof-apply-a-random-subset-of-transforms)
2. fix logging so that `mode` is used to decide where the log data is being saved. 
3. Fix mosaic implementation. See the albumentation functionality for doing the same thing but with randomizing the distribution or placement of individual images and differences in selecting what images to include [link](https://albumentations.ai/docs/examples/example-mosaic/)
4. Adding better descriptions to the `collate.py` docstrings.
5. Fix dataloader visualiztion (showing 16 repeating images when it should only be showing 4 unique images).

**Comment by sbaghba** [comment]
batch_visualization_mask
batch_visualization_image

**Comment by mkutu** [comment]
@sbaghba my comments and requests are mostly format-based. I let us go outside our coding standard for too long and now it's catching up to bite us. We need to make some structural changes.

We should be able to run viz_augments stand-alone, without having to run train pipeline. We want to be able to do this because it gives us some time to review the augments and adjust as needed before we train. Having it also run when the training pipeline is called (what is currently implemented) is good because it serves as a sanity check, be we don't want to have the entire train pipeline be running while we review augments, stop the heavy process, make changes, then start it again, and accumulate a bunch of unnecessary stop and starts plus log data.  

This is why I wanted the train.py to have a task registry early on. Adding these "tasks" is easy once that's setup.

Correct this branch by: 
1. - [x]  create `src/train_utils` and move `src/utils/augmentation_vizualizer.py`, `src/utils/gpu_utils.py` , `src/models`, and `src/data` into this new "mode" location. You can leave the others in `src/utils` since they're being used in more than 1 "mode". 
2. - [x] Remove `src/utils/visualize_dataloader.py` if you're not using it. 
3. - [x]  create `train_pipeline.py` to the `train_utils` that has the bulk of your `train.py` content in it
4. - [x] Refactor train.py to have a task registry that, for now, only has 2 tasks, `viz_augments` and `train`. Set it up like `mask_gen.py`. 
5. - [x] Update main configs 
```yaml
  train:
    vis_augment: true
    train_pipeline: true
```
6. - [x]  Refactor the `viz_augment` and `train` tasks to only take in the config (`cfg`). Review @navjot-nangia `src/mask_gen_utils/mask_inspect.py` `main()` function as reference. 
7. - [x]  Save the output plots in `projects/{project.name}/train/augments/version_{n}` along with a copy of the hydra configs that were used (for reproducibility) 
8. - [x]  Address #100 
9. - [x] Minor cleaning up of scripts

This may seem like a lot but is mostly just organizing. Shouldn't take long. Let's try and have it done for review by tomorrow for our checkin.

**Comment by sbaghba** [comment]
added my own implementation inspired by the method you showed. didnt change dataset.py so minimal changes required and it shows the augmentations fine.

**Comment by sbaghba** [comment]
batch_visualization_image
batch_visualization_mask

with image norm

**Comment by sbaghba** [comment]
batch_visualization_image
batch_visualization_mask

with channel norm

**Comment by sbaghba** [comment]
the PR is done

**Comment by mkutu** [comment]
@sbaghba 
- [x] During training, Wandb only creates the batch image, all other plots like training and validation plots are missing. Even when adding an entity in cfg.train.logger.entity (**high priority**)
- [x] FDA implementation needs a reference image. As of now, no augmentation is created when this is enabled. Add a warning or raise an error that says "not implemented" or something like that (low priority)
- [x] HistogramMatching fails with error. This also needs a reference image and likely more. Add a warning or raise an error that says "not implemented" or something like that (low priority)

Notes: 
I made some changes including:
1. adjusted some augmentation parameters
2. added batch viz and augment viz when `cfg.train.dataloader_visualizer.enabled=True`
    1. I also fixed the fact that you weren't using `cfg.train.dataloader_visualizer.enabled` just the presence of `cfg.train.dataloader_visualizer` to trigger the vizualizer.

**Review by mkutu — CHANGES_REQUESTED** [review]
See https://github.com/precision-sustainable-ag/Field-SegmentationTraining/pull/89#issuecomment-3228924057
